ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -1716,3 +1713,52 @@ enum pkcs11_rc set_key_data(struct obj_attrs **head, void *data,
 		return PKCS11_CKR_GENERAL_ERROR;
 	}
 }
+
+enum pkcs11_rc add_missing_attribute_id(struct obj_attrs **pub_head,
+					struct obj_attrs **priv_head)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	void *id1 = NULL;
+	uint32_t id1_size = 0;
+	void *id2 = NULL;
+	uint32_t id2_size = 0;
+
+	assert(pub_head);
+	assert(priv_head);
+
+	rc = get_attribute_ptr(*pub_head, PKCS11_CKA_ID, &id1, &id1_size);
+	if (rc) {
+		if (rc != PKCS11_RV_NOT_FOUND)
+			return rc;
+		id1 = NULL;
+	} else if (!id1_size) {
+		id1 = NULL;
+	}
+
+	rc = get_attribute_ptr(*priv_head, PKCS11_CKA_ID, &id2, &id2_size);
+	if (rc) {
+		if (rc != PKCS11_RV_NOT_FOUND)
+			return rc;
+		id2 = NULL;
+	} else if (!id2_size) {
+		id2 = NULL;
+	}
+
+	/* Both have value -- let them be what caller has specified them */
+	if (id1 && id2)
+		return PKCS11_CKR_OK;
+
+	/* Both are empty -- leave empty values */
+	if (!id1 && !id2)
+		return PKCS11_CKR_OK;
+
+	/* Cross copy CKA_ID value */
+	if (id1)
+		return set_attribute(priv_head, PKCS11_CKA_ID,
+				     id1, id1_size);
+	else
+		return set_attribute(pub_head, PKCS11_CKA_ID,
+				     id2, id2_size);
+
+	return PKCS11_CKR_OK;
```

Comment:
can remove this instruction.

by the way, lines 1757 and 1760 fit in single lines:
```c
	if (id1)
		return set_attribute(priv_head, PKCS11_CKA_ID, id1, id1_size);
	else
		return set_attribute(pub_head, PKCS11_CKA_ID, id2, id2_size)
```

Reply:
Right.